src/DurableTask.Netherite/TransportLayer/EventHubs/EventHubsProcessor.cs

```diff
@@ -428,59 +453,54 @@ async Task IEventProcessor.ProcessEventsAsync(PartitionContext context, IEnumera
                 {
                     this.traceHelper.LogDebug("EventHubsProcessor {eventHubName}/{eventHubPartition}({incarnation}) is processing packets", this.eventHubName, this.eventHubPartition, current.Incarnation);
 
-                    // we need to update the next expected seqno even if the iterator returns nothing, since it may have discarded some packets.
-                    // iterators do not support ref arguments, so we use a simple wrapper class to work around this limitation
-                    MutableLong nextPacketToReceive = new MutableLong() { Value = current.NextPacketToReceive.seqNo };
+                    // we need to update the next expected position tuple (seqno,batchpos) even if the iterator returns nothing, since it may have discarded some packets.
+                    // iterators do not support ref arguments, so we define a Position object with mutable fields to work around this limitation
+                    Position nextPacketToReceive = new Position() { SeqNo = current.NextPacketToReceive.seqNo, BatchPos = current.NextPacketToReceive.batchPos };
 
-                    await foreach ((EventData eventData, PartitionEvent[] events, long seqNo) in this.blobBatchReceiver.ReceiveEventsAsync(this.taskHubGuid, packets, this.shutdownToken, nextPacketToReceive))
+                    await foreach ((EventData eventData, PartitionEvent[] events, long seqNo, BlockBlobClient blob) in this.blobBatchReceiver.ReceiveEventsAsync(this.taskHubGuid, packets, this.shutdownToken, current.ErrorHandler, nextPacketToReceive))
                     {
                         for (int i = 0; i < events.Length; i++)
                         {
                             PartitionEvent evt = events[i];
 
+                            if (evt == null)
+                            {
+                                continue; // was skipped over by the batch receiver because it is already processed
+                            }
+                            
                             if (i < events.Length - 1)
                             {
                                 evt.NextInputQueuePosition = seqNo;
                                 evt.NextInputQueueBatchPosition = i + 1;
+                                this.pendingDelivery.Enqueue(new EventEntry(seqNo, i, evt));
                             }
                             else
                             {
                                 evt.NextInputQueuePosition = seqNo + 1;
+                                evt.NextInputQueueBatchPosition = 0;
+                                this.pendingDelivery.Enqueue(new EventEntry(seqNo, i, evt, long.Parse(eventData.SystemProperties.Offset), blob));
                             }
```

Comment:
what scenarios do these two cases (the if and the else) represent? Trying to understand how they're distinct, as well as why we didn't have this enqueue'ing in the past.

Reply:
The point of the pendingDelivery queue was originally just to redeliver events to new incarnations of a partition if necessary. However, I am now also using it to manage the lifecycle of blob batches.

Stuff gets removed from that queue when a partition has persisted the delivery of an event. Originally I was storing entire batches in that queue. But it turned out that manage the lifecycle of these "pending delivery" events at the batch level does not work well (and I had several bugs in that code) because tracking when an event gets persisted depends on the type of event (read vs update), so now this is being tracked at the individual event level.

So instead of adding the entire batch to this queue (line 460 of the original file) we are now adding individual events (snippet above). The two clauses deal with the difference between an event in the middle of the batch, and the last event of the batch.